Add support for alias parameter per #169

elipousson opened the pull request:
## Checklist 

- [X] update NEWS.md
- [X] documentation updated with `devtools::document()`
- [X] `devtools::check()` passes locally

## Changes 

Add support for alias parameter per #169 and also:

- adds `pull_field_aliases()` helper function
- expands tests to cover new `arc_read()` features/checks
- adds informative message if n_feats > n_max (needed due to low default `n_max` value for `arc_read()`)

This is a follow-up on this draft PR: https://github.com/R-ArcGIS/arcgislayers/pull/167

I moved `arc_read()` ahead of the changes to `filter_geom` handling since we still had some uncertainty around the best way to handle the latter.

**Issues that this closes** 

This one is fixed but I'm not sure if it is from this PR or the last PR: https://github.com/R-ArcGIS/arcgislayers/issues/170

Also fixes: https://github.com/R-ArcGIS/arcgislayers/issues/169

## Follow up tasks

The remaining tasks included in the original mega-PR.

JosiahParry commented on R/arc-select.R:
```diff
@@ -224,26 +233,27 @@ collect_layer <- function(
-  if (is.null(n_feats)) {
-    cli::cli_abort(c(
-      "Cannot determine the number of features in request.",
-      "i" = "did you set custom parameters via {.arg ...}?"
```
this message is really important because dots is used to pass in extra arguments that might not be exposed via function arguments _or_ it might have captured mistyped names in the arguments. I would like to keep some variety of this

JosiahParry commented on R/arc-select.R:
```diff
@@ -224,26 +233,27 @@ collect_layer <- function(
-      "Cannot determine the number of features in request.",
-      "i" = "did you set custom parameters via {.arg ...}?"
-    ), call = error_call)
+  if (is.null(n_feats) && is.null(n_max)) {
```
The `count_result()` function is used to determine how many pages to create and iterate through. When the `count_results()` fails, that implies that the subsequent queries will also fail. It will return a NULL then the check for `n_feats > n_max` will evaluate to `logical(0)` which causes an error. So i think this check _needs_ to be for just `is.null(n_feats)`

JosiahParry replied:
Current PR behavior 
```r
furl <- "https://services3.arcgis.com/ZvidGQkLaDJxRSJ2/arcgis/rest/services/PLACES_LocalData_for_BetterHealth/FeatureServer/0"

arc_open(furl) |> 
  arc_select(spatialRel = "wouldnt you like to know?", n_max = 10)
```

<img width="631" alt="image" src="https://github.com/R-ArcGIS/arcgislayers/assets/13534508/b9d1dd19-0658-4ad3-974a-cdb5563e299e">

After removing the second check: 
```suggestion
  if (is.null(n_feats)) {
```

<img width="721" alt="image" src="https://github.com/R-ArcGIS/arcgislayers/assets/13534508/2a3ed196-9d62-4b1c-8e68-e4befa041ab9">

JosiahParry commented on R/arc-select.R:
```diff
@@ -224,26 +233,27 @@ collect_layer <- function(
-        "*" = "Check to make sure your {.arg where} statement is valid."
-      ),
-      call = error_call
+    cli::cli_bullets(
+      c(
+        "i" = "Query results limited to {n_max} out of {n_feats} available feature{?s}.",
+        "!" = "Increase {.arg n_max} value to return all selected features."
+        )
```
this is a verbose message which doesn't have any way to suppress the output. If we want to be extra loud, we should have a way to get rid of it. In `arcgisgeocode` I've added `.progress` arguments to include progress bars from `httr2::req_perform_parallel()` perhaps here we can use a `verbose = TRUE` default argument then conditionally provide this message.

JosiahParry commented on R/arc-read.R:
```diff
@@ -13,10 +17,13 @@
-#'   `"alias"`, names are set to match the alias names for the layer, if available.
+#'   the existing sf column name is retained.
 #' @param col_select Default `NULL`. A character vector of the field names to be
 #'   returned. By default, all fields are returned.
+#' @param alias Default "drop". Supported options: `c("drop", "label",
+#'   "replace")`. If "drop", field alias values are ignored. If "label", field
+#'   alias values are assigned as a label attribute for each field. If "replace"
+#'   and col_names is `TRUE`, field alias values are used as the column names.
```
```suggestion
#'   returned. By default, all fields are returned.
#' @param alias Default "drop". Supported options: `c("drop", "label",
#'   "replace")`. If `"drop"`, field alias values are ignored. If `"label"`, field
#'   alias values are assigned as a label attribute for each field. If `"replace"`
#'   and col_names is `TRUE`, field alias values are used as the column names.
```

JosiahParry commented on R/arc-read.R:
```diff
@@ -13,10 +17,13 @@
+#'   the existing sf column name is retained.
 #' @param col_select Default `NULL`. A character vector of the field names to be
 #'   returned. By default, all fields are returned.
+#' @param alias Default "drop". Supported options: `c("drop", "label",
```
```suggestion
#' @param alias Default `"drop"`. Supported options: `c("drop", "label",
```

JosiahParry commented on R/arc-read.R:
```diff
@@ -68,109 +75,175 @@ arc_read <- function(
   # Use existing names by default
-  nm <- layer_nm
-  sf_column_nm <- attr(x, "sf_column")
+  existing_nm <- names(layer)
```
Use colnames just in case something weird happens and row-names exist
```suggestion
  existing_nm <- colnames(layer)
```

JosiahParry commented on R/arc-read.R:
```diff
@@ -68,109 +75,175 @@ arc_read <- function(
+    )
+  }
+
+  if (rlang::is_true(col_names) && alias == "drop" && !has_name_repair) {
+    return(layer)
+  }
```
can we add a comment here that documents what this logic is?

elipousson replied:
Definitely can add a comment or could drop this entirely. This is just an early return that skips all of the subsequent code if none of it really applies. I'm generally a fan of early returns where possible but it is a bit of a messy mix of conditions.

JosiahParry replied:
early return is totally fine! I'm just not great at holding more than two conditions in my head lol

JosiahParry commented on R/arc-select.R:
```diff
@@ -224,26 +233,27 @@ collect_layer <- function(
+  if (is.null(n_feats) && is.null(n_max)) {
+    cli::cli_abort(
+      c("Can't determine the number of features for {.arg x}.",
+        "*" = "Check to make sure your {.arg where} statement is valid or
+        set a value for {.arg n_max}."
+      ),
```
```suggestion
        "*" = "Check to make sure your {.arg where} statement is valid or any custom parameters set via {.arg ...}."
      ),
```

JosiahParry commented:
Thank you for the PR! There are a couple of things that we need to clean up: 

1. the check for `is.null(n_max)` 
2. add a test for when count_results() will return 0: `arc_select(arc_open(furl), spatialRel = "wouldnt you like to know?", n_max = 10)`
3. fix the error message when `is.null(n_feats)` to indicate that the problem can be coming from `...` 
4. I think we should remove the `cli_bullets()` for now as introducing a verbose argument in this PR is out of scope. 

I'm happy to make these changes my self if you are okay with that!

elipousson commented:
I'm happy to handle the modifications to the check – I think I migrated that whole block of code to a helper check function in the big PR and would like to do the same here if that works for you. Very happy for you to add the tests so I know when I've got it right.

If we don't include the `cli_bullets()` message in this PR, how do you feel about increasing the default `n_max` value to 1e+06 or 1e+07. As is, I think it is a pretty critical alert (and it can be suppressed using a cli setting) that users may miss otherwise. Alternatively, I can move it into another PR but I was hoping to finish the additions to arc_read with this PR so we can move on to the `filter_geom` fix (which I think I've got figured out).

On the topic of tests and the overly verbose messages – have you thought about adding withr to Suggests and using it within the tests? The way the tests currently handle options could be migrated to withr with ease and there are some benefits: https://testthat.r-lib.org/articles/test-fixtures.html#test-fixtures.

JosiahParry commented:
We can smash `n_max = Inf` for `arc_read()` if you want. I've encountered a fair few datasets now where even 50 obs might crush the machine! But alas, if people want it all, they want it all. 

I do think the `cli_bullets()` is handy. But I think relying on a global cli option is a bit awkward. I think it would be a good candidate for a follow up PR !

elipousson commented:
Took a little longer than expected to get everything working as expected but I think this now addresses all of your feedback and integrates more (but not all!) of the changes from that draft PR. In addition to the changes to `arc_read()`, the changes include reworking the parameter validation logic and adding some helper functions to make the code within `collect_layer()` a little easier to follow.

JosiahParry reviewed (approved):
🚀 🎉
